### PR TITLE
Add a debug implementation for Address

### LINKED ISSUE #54: Could `Address` implement `fmt::Debug`?
Sometimes it is useful for messages to implement `fmt::Debug` but if they contain `Address`es, that is not possible without a manual impl. Could `Address` implement `Debug`?

Not sure if we can print something useful but the counter value could be interesting for example!

### PR DESCRIPTION
**Pull request by Restioson**
This does not address MessageChannel as I just haven't looked into it yet. I'll look into it before this gets merged, though. Closes #54.

- [x] Implement Debug on RefCounter and show reference counts
- [x] Implement for MessageChannel
- [x] Add test

### CODE REVIEW

**Review by thomaseizinger — COMMENTED** [review]
I think that is a good starting point. Implementing Debug on RefCounter could also be interesting to see the actual refcount.

**Comment by Restioson** [comment]
That makes sense

**Comment by Restioson** [comment]
Is this what you were thinking of when it comes to the reference count debugging @thomaseizinger ?

**Comment by thomaseizinger** [comment]
Yeah this looks good! Perhaps we can add a test that asserts the Debug implementation? This would make it even easier to review and we make sure it is not broken :)

File: src/address.rs
Comment (thomaseizinger) [inline src/address.rs:53]: The strong/weak part is duplicated here, isn't it? I think we can remove it from the `Address` type itself and only have it as part of the `ref_counter` field?

Comment (Restioson) [reply src/address.rs:53]: That is technically true. I mostly kept it to show the full type since  Address<A> might seem to imply strong as that's the default

**Comment by Restioson** [comment]
Sounds like a good plan

File: src/address.rs
Comment (thomaseizinger) [reply src/address.rs:53]: Fair enough!
Let's see what it looks like when printed but it would be nice to reduce the duplication one way or another :)

**Comment by Restioson** [comment]
I've added a test for this now. One remaining unresolved question is whether to report the weak count strictly accurately (i.e number of weak references, incl. the one kept inside Context), or to make it count the number of weak addresses (statically one less than the strict count).

**Comment by thomaseizinger** [comment]
> I've added a test for this now. One remaining unresolved question is whether to report the weak count strictly accurately (i.e number of weak references, incl. the one kept inside Context), or to make it count the number of weak addresses (statically one less than the strict count).

I think reporting the actual number of addresses is good enough. I see very little foot-gun potential here :)

File: tests/basic.rs
Comment (thomaseizinger) [inline tests/basic.rs:278]: Technically, this doesn't need to be an `async` test.

File: tests/basic.rs
Comment (thomaseizinger) [inline tests/basic.rs:287]: If you use raw strings (`r#""#`), you can omit the escaping of the newlines.

Comment (thomaseizinger) [reply tests/basic.rs:287]: I just realized that this isn't even a multiline string. Is this just on two lines because of the formatter?

File: tests/basic.rs
Comment (thomaseizinger) [inline tests/basic.rs:295]: Seeing this, I'd suggest to simplify the representation to:

```rust
"Address<basic::Greeter> { ref_counter: Weak { connected: true, strong_count: 2, weak_count: 2 } }"
```
or

```rust
"Address<basic::Greeter, xtra::refcount::Weak> { connected: true, strong_count: 2, weak_count: 2 }"
```

File: tests/basic.rs
Comment (Restioson) [reply tests/basic.rs:278]: Good point

File: tests/basic.rs
Comment (Restioson) [reply tests/basic.rs:287]: It's slightly more readable broken across two lines, otherwise it tends to go off the monitor

File: tests/basic.rs
Comment (Restioson) [reply tests/basic.rs:295]: This makes sense. I think I prefer the first probably since it'll show which variant of ref counter is used if it's erased through Either

File: tests/basic.rs
Comment (thomaseizinger) [inline tests/basic.rs:279]: You could avoid the long type hint with turbo-fish syntax but that is a style preference :)


```suggestion
    let (addr1, _ctx) = Context::<Greeter>::new(None);
```

**Review by thomaseizinger — APPROVED** [review]
Implementation looks good to me :)

File: examples/basic_wasm_bindgen/src/lib.rs
Comment (Restioson) [reply examples/basic_wasm_bindgen/src/lib.rs:20]: Yea... I had thought that I'd already rebased it on master but apparently not

File: src/refcount.rs
Comment (Restioson) [reply src/refcount.rs:48]: That's true - especially since it's only used in a debug implementation and doesn't require upgrading, downgrading, or dropping anything

**Comment by Restioson** [comment]
Going to go ahead and merge this now if CI passes

**Comment by Restioson** [comment]
I'm pretty confident that CI will pass once #80 lands, since the current failure has nothing to do with any code changes in this branch, so I am going to go ahead and merge this now.

**Pull request merged**
